hedera-node/hedera-app/src/main/java/com/hedera/node/app/config/ConfigProviderImpl.java

```diff
@@ -0,0 +1,71 @@
+/*
+ * Copyright (C) 2023 Hedera Hashgraph, LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hedera.node.app.config;
+
+import com.hedera.node.app.config.internal.ConfigurationAdaptor;
+import com.hedera.node.app.config.internal.VersionedConfigImpl;
+import com.hedera.node.app.service.mono.context.properties.GlobalDynamicProperties;
+import com.hedera.node.app.service.mono.context.properties.PropertySource;
+import com.hedera.node.app.spi.config.ConfigProvider;
+import com.hedera.node.app.spi.config.VersionedConfiguration;
+import com.swirlds.common.threading.locks.AutoClosableLock;
+import com.swirlds.common.threading.locks.Locks;
+import com.swirlds.config.api.Configuration;
+import edu.umd.cs.findbugs.annotations.NonNull;
+import java.util.Objects;
+import java.util.concurrent.atomic.AtomicReference;
+
+/**
+ * Implementation of the {@link ConfigProvider} interface.
+ */
+public class ConfigProviderImpl implements ConfigProvider {
+
+    private final PropertySource propertySource;
+
+    private final AtomicReference<VersionedConfiguration> configuration;
+
+    private final AutoClosableLock updateLock = Locks.createAutoLock();
+
+    /**
+     * Constructor.
+     *
+     * @param propertySource the property source to use. All properties will be read from this source.
+     */
+    public ConfigProviderImpl(@NonNull final PropertySource propertySource) {
+        this.propertySource = Objects.requireNonNull(propertySource, "propertySource");
+        final Configuration config = new ConfigurationAdaptor(propertySource);
+        configuration = new AtomicReference<>(new VersionedConfigImpl(config, 0));
+    }
+
+    /**
+     * This method must be called if a property has changed. It will update the configuration and increase the version.
+     * This should happen whenever {@link GlobalDynamicProperties#reload()} is called.
+     */
+    public void update() {
+        try (final var lock = updateLock.lock()) {
+            final Configuration config = new ConfigurationAdaptor(propertySource);
+            final VersionedConfiguration versionedConfig =
+                    new VersionedConfigImpl(config, this.configuration.get().getVersion() + 1);
+            configuration.set(versionedConfig);
```

Comment:
Atomic is used that a parallel call of `getConfiguration()` does not return a "broken" config.
The lock is used to not allow 2 parallel calls of the update method, this would destroy the version number.

Only using lock would work if I add it to `getConfiguration()`, too. But that would look the getter and I want to avoid it. As already mentioned it might be good enough to use volatile for the config instead of an atomic reference since no changes happen after config has been written but for now I would like to keep it as it is since the PR should be merged in near future.

Reply:
(I don't see how it could return a "broken" config.  It's either going to return the old one, or the new one.  Maybe you can't predict _which_ one in the case of "simultaneous" get + set or two "simultaneous" sets.  But it's going to be one or the other.  Reason I think this is that the `versionedConfig` isn't updated in place.  You create a new one.)

(The _user_ in trying to get _two_ fields "at the same time" might call `getConfiguration` twice and get two different configurations.  But that's on him, and the atomic reference doesn't prevent that either.)

(Ah, the issue is that multiple threads reading the config - without volatile or atomic one thread might continue to read the old configuration for some period of time.  So it's not a "broken" config that's the problem: It's an out-of-date config due to caching in the thread.)

Reply:
The lock is needed because the current config is used to increment the version. However, I'd use volatile instead of AtomicReference to store the configuration in this case since you are not doing a compare and swap.

Reply:
One small annoyance: on the previous line, you reference `this.configuration` and on this line you use `configuration`.

Reply:
Not saying we should do this but, if we want to encapsulate version logic, `VersionedConfigImpl` could also take a set of updates and return a new config with an incremented version.

Reply:
good comments. Will resolve them for now to finally merge the PR. Will create another PR to address them.